mm2src/coins/z_coin/z_rpc.rs

```diff
@@ -177,31 +358,79 @@ pub(super) async fn init_light_client(
             Ok(db)
         }
     })
-    .await?;
+    .await
+}
+
+pub(super) async fn init_light_client(
+    _lightwalletd_urls: Vec<String>,
+    lightwalletd_url: Uri,
+    cache_db_path: PathBuf,
+    wallet_db_path: PathBuf,
+    consensus_params: ZcoinConsensusParams,
+    check_point_block: Option<CheckPointBlockInfo>,
+    evk: ExtendedFullViewingKey,
+) -> Result<(AsyncMutex<SaplingSyncConnector>, WalletDbShared), MmError<ZcoinClientInitError>> {
+    let blocks_db =
+        async_blocking(|| BlockDb::for_path(cache_db_path).map_to_mm(ZcoinClientInitError::BlocksDbInitFailure))
+            .await?;
+
+    let wallet_db = create_wallet_db(wallet_db_path, consensus_params.clone(), check_point_block, evk).await?;
+
+    let (sync_status_notifier, sync_watcher) = channel(1);
+    let (on_tx_gen_notifier, on_tx_gen_watcher) = channel(1);
+
+    let wallet_db = Arc::new(Mutex::new(wallet_db));
+    let sync_handle = SaplingSyncLoopHandle {
+        current_block: BlockHeight::from_u32(0),
+        blocks_db: Mutex::new(blocks_db),
+        wallet_db: wallet_db.clone(),
+        consensus_params,
+        sync_status_notifier,
+        on_tx_gen_watcher,
+        watch_for_tx: None,
+    };
 
     let tonic_channel = Channel::builder(lightwalletd_url)
         .tls_config(ClientTlsConfig::new())
-        .map_to_mm(ZcoinLightClientInitError::TlsConfigFailure)?
+        .map_to_mm(ZcoinClientInitError::TlsConfigFailure)?
         .connect()
         .await
-        .map_to_mm(ZcoinLightClientInitError::ConnectionFailure)?;
-    let grpc_client = CompactTxStreamerClient::new(tonic_channel);
+        .map_to_mm(ZcoinClientInitError::ConnectionFailure)?;
+    let rpc_copy = CompactTxStreamerClient::new(tonic_channel);
+    let abort_handle = spawn_abortable(light_wallet_db_sync_loop(sync_handle, Box::new(rpc_copy)));
+
+    Ok((
+        SaplingSyncConnector::new_mutex_wrapped(sync_watcher, on_tx_gen_notifier, abort_handle),
+        wallet_db,
+    ))
+}
 
+pub(super) async fn init_native_client(
+    native_client: NativeClient,
+    cache_db_path: PathBuf,
+    wallet_db_path: PathBuf,
+    consensus_params: ZcoinConsensusParams,
+    check_point_block: Option<CheckPointBlockInfo>,
+    evk: ExtendedFullViewingKey,
+) -> Result<(AsyncMutex<SaplingSyncConnector>, WalletDbShared), MmError<ZcoinClientInitError>> {
+    let blocks_db =
```

Comment:
Databases' initialization code is duplicated among `init_native_client` and `init_light_client`. It seems you can move it to `ZCoinBuilder::build` instead.

Reply:
done